File: backend/app/schemas/policy.py
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict


PolicyCategory = Literal[
    "housing",
    "finance",
    "welfare",
    "employment",
    "startup",
    "education",
    "other",
]
ApplicationSchedule = Literal[
    "fixed_period",
    "always",
    "until_budget_exhausted",
]
ApplicationStatus = Literal["open", "closed", "scheduled"]
DataQualityStatus = Literal["valid", "partial"]
EligibilityCoverage = Literal["complete", "partial", "unknown"]
EligibilityCategory = Literal[
    "age",
    "region",
    "income",
    "asset",
    "employment",
    "education",
    "housing",
    "household",
    "other",
]
EvidenceLocatorType = Literal["source_field", "css_selector"]
InstitutionalContactKind = Literal["phone", "official_channel"]


class EligibilityEvidenceRead(BaseModel):
    source_id: str
    source_url: str
    collected_at: datetime
    locator_type: EvidenceLocatorType
    locator: str


class EligibilityConditionRead(BaseModel):
    category: EligibilityCategory
    text: str
    evidence: list[EligibilityEvidenceRead]


class EligibilityDocumentRead(BaseModel):
    text: str
    evidence: list[EligibilityEvidenceRead]


class InstitutionalContactRead(BaseModel):
    kind: InstitutionalContactKind
    label: str
    value: str
    evidence: list[EligibilityEvidenceRead]


class EligibilitySummaryRead(BaseModel):
    coverage: EligibilityCoverage
    requirements: list[EligibilityConditionRead]
    exclusions: list[EligibilityConditionRead]
    preferences: list[EligibilityConditionRead]
    documents: list[EligibilityDocumentRead]
    unknowns: list[EligibilityConditionRead]
    institutional_contacts: list[InstitutionalContactRead]


class PolicyBase(BaseModel):
    schema_version: str = "1.0.0"
    source_id: str
    source_name: str
    external_id: str | None = None
    title: str
    organization: str | None = None
    summary: str | None = None
    category_text: str | None = None
    categories: list[PolicyCategory]
    application_period_text: str | None = None
    application_start: date | None = None
    application_end: date | None = None
    application_schedule: ApplicationSchedule | None = None
    application_status: ApplicationStatus | None = None
    region_text: str | None = None
    regions: list[str]
    age_min: int | None = None
    age_max: int | None = None
    age_condition_text: str | None = None
    eligibility_text: str | None = None
    support_content: str | None = None
    application_method: str | None = None
    education_statuses: list[str]
    employment_statuses: list[str]
    required_conditions: list[str]
    preferred_conditions: list[str]
    excluded_conditions: list[str]
    source_url: str
    collected_at: datetime
    data_quality_status: DataQualityStatus


class PolicyRead(PolicyBase):
    """Public policy response. Raw provenance is intentionally excluded."""

    id: int
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class PolicyDetailRead(PolicyRead):
    """Detail response with source-backed eligibility evidence."""

    eligibility_summary: EligibilitySummaryRead


class PolicyListResponse(BaseModel):
    total: int
    page: int
    limit: int
    items: list[PolicyRead]

File: frontend/src/types/policy.ts
export type PolicyCategory =
  | 'housing'
  | 'finance'
  | 'welfare'
  | 'employment'
  | 'startup'
  | 'education'
  | 'other';

export type ApplicationSchedule =
  | 'fixed_period'
  | 'always'
  | 'until_budget_exhausted';

export type ApplicationStatus = 'open' | 'closed' | 'scheduled';

export type PublicDataQualityStatus = 'valid' | 'partial';

export type PolicySort =
  | 'default'
  | 'title_asc'
  | 'title_desc'
  | 'deadline_asc'
  | 'deadline_desc'
  | 'collected_desc'
  | 'collected_asc';

export type EligibilityCoverage = 'complete' | 'partial' | 'unknown';

export type EligibilityCategory =
  | 'age'
  | 'region'
  | 'income'
  | 'asset'
  | 'employment'
  | 'education'
  | 'housing'
  | 'household'
  | 'other';

export interface EligibilityEvidenceDto {
  source_id: string;
  source_url: string;
  collected_at: string;
  locator_type: 'source_field' | 'css_selector';
  locator: string;
}

export interface EligibilityConditionDto {
  category: EligibilityCategory;
  text: string;
  evidence: EligibilityEvidenceDto[];
}

export interface EligibilityDocumentDto {
  text: string;
  evidence: EligibilityEvidenceDto[];
}

export interface InstitutionalContactDto {
  kind: 'phone' | 'official_channel';
  label: string;
  value: string;
  evidence: EligibilityEvidenceDto[];
}

export interface EligibilitySummaryDto {
  coverage: EligibilityCoverage;
  requirements: EligibilityConditionDto[];
  exclusions: EligibilityConditionDto[];
  preferences: EligibilityConditionDto[];
  documents: EligibilityDocumentDto[];
  unknowns: EligibilityConditionDto[];
  institutional_contacts: InstitutionalContactDto[];
}

export interface PolicyDto {
  schema_version: '1.0.0' | '1.1.0' | '1.2.0';
  source_id: string;
  source_name: string;
  external_id: string | null;
  title: string;
  organization: string | null;
  summary: string | null;
  category_text: string | null;
  categories: PolicyCategory[];
  application_period_text: string | null;
  application_start: string | null;
  application_end: string | null;
  application_schedule: ApplicationSchedule | null;
  application_status: ApplicationStatus | null;
  region_text: string | null;
  regions: string[];
  age_min: number | null;
  age_max: number | null;
  age_condition_text: string | null;
  eligibility_text: string | null;
  support_content: string | null;
  application_method: string | null;
  education_statuses: string[];
  employment_statuses: string[];
  required_conditions: string[];
  preferred_conditions: string[];
  excluded_conditions: string[];
  source_url: string;
  collected_at: string;
  data_quality_status: PublicDataQualityStatus;
  id: number;
  created_at: string;
  updated_at: string;
}

export interface PolicyDetailDto extends PolicyDto {
  eligibility_summary: EligibilitySummaryDto;
}

export interface PolicyListResponse {
  total: number;
  page: number;
  limit: number;
  items: PolicyDto[];
}

export interface PolicyListQuery {
  page?: number;
  limit?: number;
  category?: PolicyCategory;
  region?: string;
  status?: ApplicationStatus;
  include_partial?: boolean;
  sort?: PolicySort;
}
